File: app/src/main/java/io/scribeapp/classifier/model/GestureClassificationRequest.java
package io.scribeapp.classifier.model;

import android.gesture.Gesture;
import android.gesture.GestureStroke;
import android.graphics.Bitmap;
import android.util.Log;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.io.ByteArrayOutputStream;
import java.lang.reflect.Field;

import io.scribeapp.classifier.utils.Utils;

public class GestureClassificationRequest extends ClassificationRequest {
    public Gesture gesture;

    public GestureClassificationRequest(Gesture gesture) {
        this.gesture = gesture;
    }

    public byte[] toByteArray() {
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        Utils.getBitmapFromGesture(gesture).compress(Bitmap.CompressFormat.PNG, 100, os);
        return os.toByteArray();
    }

    @Override
    public JsonElement toJson() {
        JsonArray strokes = new JsonArray();
        for (GestureStroke stroke : gesture.getStrokes()) {
            JsonArray pointArray = new JsonArray();
            Field f = null;
            long[] timestamps = null;
            try {
                f = stroke.getClass().getDeclaredField("timestamps");
                f.setAccessible(true);
                timestamps = (long[]) f.get(stroke);
            } catch (NoSuchFieldException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }

            for (int i = 0; i < stroke.points.length; i += 2) {
                JsonObject point = new JsonObject();
                point.addProperty("x", stroke.points[i]);
                point.addProperty("y", stroke.points[i + 1]);
                if (timestamps != null) {
                    point.addProperty("time", timestamps[i / 2]);
                }
                pointArray.add(point);
            }
            strokes.add(pointArray);
        }
        return strokes;
    }
}

File: app/src/main/kotlin/io/scribe/classifier/model/GestureClassificationRequest.kt
package io.scribe.classifier.model

import android.gesture.Gesture
import java.io.ByteArrayOutputStream
import io.scribe.classifier.utils.Utils
import android.graphics.Bitmap
import com.google.gson.JsonElement
import com.google.gson.JsonArray
import java.lang.reflect.Field
import com.google.gson.JsonObject
import android.gesture.GestureStroke

public class GestureClassificationRequest(public var gesture: Gesture) : ClassificationRequest {

    override fun toByteArray(): ByteArray {
        val os = ByteArrayOutputStream()
        Utils.getBitmapFromGesture(gesture).compress(Bitmap.CompressFormat.PNG, 100, os)
        return os.toByteArray()
    }

    override fun toJson(): JsonElement {
        val strokes = JsonArray()
        for (stroke in gesture.getStrokes()) {
            val pointArray = JsonArray()

            val timestamps = extractTimestamp(stroke)

            for (i in 0..stroke.points.size() - 1 step 2) {
                val point = JsonObject()
                point.addProperty("x", stroke.points[i])
                point.addProperty("y", stroke.points[i + 1])
                point.addProperty("time", timestamps[i / 2])
                pointArray.add(point)
            }
            strokes.add(pointArray)
        }
        return strokes
    }

    private fun extractTimestamp(stroke: GestureStroke): LongArray {
        var f: Field?
        var timestamps = LongArray(0)
        try {
            f = stroke.javaClass.getDeclaredField("timestamps")
            f!!.setAccessible(true)
            timestamps = f?.get(stroke) as LongArray
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        return timestamps
    }
}